v5/core/log.go

```diff
@@ -0,0 +1,202 @@
+package core
+
+// (C) Copyright IBM Corp. 2020, 2021.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import (
+	"log"
+	"os"
+	"sync"
+)
+
+// LogLevel defines a type for logging levels
+type LogLevel int
+
+// Log level constants
+const (
+	LevelNone LogLevel = iota
+	LevelError
+	LevelWarn
+	LevelInfo
+	LevelDebug
+)
+
+// Logger is the logging interface implemented and used by the Go core library.
+// Users of the library can supply their own implementation by calling SetLogger().
+type Logger interface {
+	Log(level LogLevel, format string, inserts ...interface{})
+	Error(format string, inserts ...interface{})
+	Warn(format string, inserts ...interface{})
+	Info(format string, inserts ...interface{})
+	Debug(format string, inserts ...interface{})
+
+	SetLogLevel(level LogLevel)
+	GetLogLevel() LogLevel
+
+	IsLogLevelEnabled(level LogLevel) bool
+	IsErrorEnabled() bool
+	IsWarnEnabled() bool
+	IsInfoEnabled() bool
+	IsDebugEnabled() bool
```

Comment:
I wonder ... is there some way we can implement these methods ourselves? I mean ... as long as we have `GetLogLevel`, then the implementations of these other methods are straightforward, so why require the user to write these if they provide a custom logger?

In fact, I would go further.  We _should not allow_ the user to implement these methods, just to avoid the potential for them screwing it up.  Can we do that?

Reply:
If we were doing this in Java, we'd implement common behavior in a base class and then each implementation could subclass that.  But in Go, there's no true inheritance.  I suppose one could use composition and "inherit" some type we define with the common behaviors, but to be honest I'm not sure anyone is even implementing their own Logger's.

When making these updates I perhaps leaned a bit toward exposing more methods rather than less, to avoid another situation like we're fixing now :)
I could remove the Is[Error,Warn,Info,Debug]Enabled() methods and just leave IsLogLevelEnabled(LogLevel) there.

As for your last question, I don't know off-hand how we'd prevent someone from implementing specific methods, but if someone is implementing their own Logger they'd need to implement all the methods of the interface, so...  maybe I didn't understand the question correctly :)

Reply:
Addressed this by removing the Is[Error,Warn,Info,Debug]Enabled() methods in favor of just IsLogLevelEnabled().